### PR TITLE
Transformers

### PR DESCRIPTION
**Pull request by wowtor**
Added various sklearn style transformers which may be useful in forensic evaluation.

### CODE REVIEW

File: lir/calibration.py [hunk 1]
```diff
@@ -165,7 +165,9 @@ def transform(self, X):
         X = X.reshape(-1, 1)
```
Comment (rolfypma-nfi) [inline lir/calibration.py:165]: re recent besluit; ophogen minor versie nummer?

Comment (wowtor) [reply lir/calibration.py:165]: ok; zoiets?

Comment (rolfypma-nfi) [reply lir/calibration.py:165]: yes

File: tests/test_lr.py [hunk 2]
```diff
@@ -43,6 +43,21 @@ def test_calculate_cllr(self):
         self.assertAlmostEqual(1.0849625007211563, metrics.cllr(*Xn_to_Xy([2], [2]*2)))
         self.assertAlmostEqual(1.6699250014423126, metrics.cllr(*Xn_to_Xy([8], [8]*8)))
 
+    def test_extreme_cllr(self):
+        self.assertEqual(np.inf, metrics.cllr(*Xn_to_Xy([np.inf, 1], [1, 1])))
```
Comment (rolfypma-nfi) [inline tests/test_lr.py:47]: is het ergens niet onlogisch dat we eerst h2, dan h1 waardes geven?

File: tests/test_lr.py [hunk 3]
```diff
@@ -43,6 +43,21 @@ def test_calculate_cllr(self):
...
+        self.assertEqual(np.inf, metrics.cllr(*Xn_to_Xy([np.inf, 0], [1, 1])))
+        self.assertEqual(np.inf, metrics.cllr(*Xn_to_Xy([1, 1], [0, 1])))
+        self.assertAlmostEqual(.5, metrics.cllr(*Xn_to_Xy([0, 0], [1, 1])))
+        self.assertAlmostEqual(.5, metrics.cllr(*Xn_to_Xy([1, 1], [np.inf, np.inf])))
+        self.assertAlmostEqual(0, metrics.cllr(*Xn_to_Xy([0, 0], [np.inf, np.inf])))
+        self.assertAlmostEqual(np.inf, metrics.cllr(*Xn_to_Xy([np.inf, np.inf], [0, 0])))
```
Comment (rolfypma-nfi) [inline tests/test_lr.py:53]: waarom hebben we hier een almost nodig, en kan bij de regel eronder wel een assert equal?

File: lir/transformers.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import numpy as np
+import sklearn
+
+
+class SumNormalizer(sklearn.base.TransformerMixin):
+    """
+    Normalize samples individually to unit sum.
+
+    See also: sklearn.preprocessing.Normalizer which normalizes to unit norm
```
Comment (rolfypma-nfi) [inline lir/transformers.py]: hoe is deze anders? wanneer wil je deze gebruiken?

File: lir/transformers.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        assert len(X.shape) == 3
+        assert X.shape[2] == 2
+
+        return np.abs(X[:,:,0] - X[:,:,1])
+
+
+class InstancePairing(sklearn.base.TransformerMixin):
+    def __init__(self, same_source_limit=None, different_source_limit=None):
+        """
+        Parameters:
```
Comment (rolfypma-nfi) [inline lir/transformers.py:39]: explanation of what this actually does and why and when you would want to use it?

File: lir/transformers.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        """
+        Expects:
+            - X is of shape (n,f) with n=number of samples; f=number of features
+            - y is of shape (n,)
+        Returns:
+            - X, y; X has shape (m,f,2); y has shape (m,); with m=number of pairs
+        Attributes:
+            - self.pairing: an array of shape (m,2) with indexes of X input array which contribute to the pairs
+        """
+        pairing = np.array(np.meshgrid(np.arange(X.shape[0]), np.arange(X.shape[0]))).T.reshape(-1, 2)  # generate all possible pairs
```
Comment (rolfypma-nfi) [inline lir/transformers.py:61]: dit kan heel erg opblazen. waarschuwing of check op lengte van X?
kan met name relevant zijn omdat als je balanced opgeeft de totalen beperkt blijven, maar dat hier pas gedaan wordt nadat alle paren gemaakt zijn.

**Review by rolfypma-nfi — APPROVED** [review]
goed om dit in te checken. het is niet meteen duidelijk wat dit doet, dus iets meer uitleg/voorbeelden lijken me erg prettig.

File: tests/test_lr.py (hunk 2 above)
Comment (wowtor) [reply tests/test_lr.py:47]: Dat is geen wijziging maar was altijd al zo. De reden is dat de resultaten op label gesorteerd zijn. Andere labels zou een optie zijn.

File: lir/transformers.py (hunk 4 above)
Comment (wowtor) [reply lir/transformers.py]: Het is een andere manier om de lengte van een vector te bepalen, namelijk als som van alle elementen. Soms lijkt dat voor de hand te liggen, bijvoorbeeld als de elementen counts van woorden zijn, en de vector een text snippet. Na normalisatie worden counts dan fracties.

File: tests/test_lr.py (hunk 2 above)
Comment (rolfypma-nfi) [reply tests/test_lr.py:47]: klopt, was meer een algemene musing. zouden labels 1 en 2 voor ons wellicht logische zijn dan 1 en 0? keer over nadenken, houdt dit PR niet tegen

File: lir/transformers.py (hunk 4 above)
Comment (rolfypma-nfi) [reply lir/transformers.py]: ik bedoelde eigenlijk: kun je aan in de doc toevoegen hoe deze anders is/zich verhoudt. Overigens is de sum van de elementen de L1 norm, (en ik vermoedt dat je bij Normalizer de L2 norm bedoelt) dus de term 'norm' is wellicht ambigu hier.

File: lir/transformers.py (hunk 5 above)
Comment (wowtor) [reply lir/transformers.py:39]: ok, fixed

File: lir/transformers.py (hunk 6 above)
Comment (wowtor) [reply lir/transformers.py:61]: Klopt, heb nu een waarschuwing toegevoegd.

File: lir/transformers.py (hunk 4 above)
Comment (wowtor) [reply lir/transformers.py]: dank je, weer wat geleerd :)

File: tests/test_lr.py (hunk 2 above)
Comment (wowtor) [reply tests/test_lr.py:47]: Ik begrijp je punt wel, maar persoonlijk vind ik het erg verwarrend als label 1 overeenkomt met value 1, en label 2 met value 0. Het generaliseer ook niet lekker, want wat als je 3 hypothesen hebt? De namen H1 en H2 zijn maar een mogelijke invulling. Een oplossing zou zijn om met de namen te blijven werken, dan kan je zelf ook de volgorde bepalen en heb je volgens mij het beste van beide. Vergt wel een aanpassing in de code hier en daar.
